### ISSUE TITLE
Socket.io for golang

### ISSUE THREAD

**Issue by rusmanplatd** [socketio/socket.io#3242]
*Note*: for support questions, please use one of these channels: [stackoverflow](http://stackoverflow.com/questions/tagged/socket.io) or [slack](https://socketio.slack.com)

### You want to:

* [ ] report a *bug*
* [x] request a *feature*

### Current behaviour
https://github.com/googollee/go-socket.io/issues/188

### Steps to reproduce (if the current behaviour is a bug)

**Note**: the best way to get a quick answer is to provide a failing test case, by forking the following [fiddle](https://github.com/darrachequesne/socket.io-fiddle) for example.

### Setup
- OS:
- browser:
- socket.io version:

**Comment by kunaldawn** [comment]
Anyone ?

**Comment by ifree92** [comment]
I also need golang client for socket.io nodejs server.
But unfortunately, googollee/go-socket.io is not working for socketio 2.0 and newer.

**Comment by odiferousmint** [comment]
*crickets*

**Comment by mofadeyunduo** [comment]
Need Golang client too, and also need Rust client.
But I can't write code using Rust. I don't know should I use socket.io because I'm not able to develop client for Golang let alone Rust . Someone can give me some advices?

**Comment by exapsy** [comment]
Yeah, I definitely would love to see Socket.IO supporting golang as well. :)

@mofadeyunduo If you're asking whether to use Socket.IO or Websockets alone, I've written an answer on StackOverflow out of my own experience with both of them.
https://stackoverflow.com/a/62848079/1712332

**Comment by gzq0616** [comment]
Need Golang client too

**Comment by alone-wolf** [comment]
Do need Golang client, too. And I'd like to contribute to it, as soon as I can.

**Comment by vinkent420** [comment]
> Yeah, I definitely would love to see Socket.IO supporting golang as well. :)
> 
> @mofadeyunduo If you're asking whether to use Socket.IO or Websockets alone, I've written an answer on StackOverflow out of my own experience with both of them. https://stackoverflow.com/a/62848079/1712332
> 
> If you have the option to make a microservice in NodeJS, or use a Socket.IO library in another language, it's definitely worth it. Websockets are very expensive to make them work on a big complex app and I had a first hand experience on trying to make them work. I mean, if all you want to do is to send back and forth a simple message without having such thing as rooms, or events, then you're good. Use Websockets, it's as simple as doing it on socket.io if not more. But if you want something more complicated, like subscriptions and isolated event rooms, using Websockets means you also have to implement all of those things from scratch and its a pain in the ass.

totally agree with bro

**Comment by pallasite99** [comment]
I tried this out and achieved a base result on my local `socket.io` server written in golang:

```golang
package main

import (
	"fmt"
	"net/http"

	"github.com/gorilla/websocket"
)

var upgrader = websocket.Upgrader{
	ReadBufferSize:  1024,
	WriteBufferSize: 1024,
}

func handleConnection(w http.ResponseWriter, r *http.Request) {
	conn, err := upgrader.Upgrade(w, r, nil)
	if err != nil {
		fmt.Println(err)
		return
	}
	defer conn.Close()

	fmt.Println("Client connected")

	for {
		messageType, p, err := conn.ReadMessage()
		if err != nil {
			fmt.Println(err)
			return
		}

		if err := conn.WriteMessage(messageType, p); err != nil {
			fmt.Println(err)
			return
		}
	}
}

func main() {
	http.HandleFunc("/ws", handleConnection)

	serverAddr := "localhost:8080"
	fmt.Printf("WebSocket server is running on ws://%s\n", serverAddr)

	err := http.ListenAndServe(serverAddr, nil)
	if err != nil {
		fmt.Println(err)
	}
}
```

This was my client:

```golang
package main

import (
	"fmt"
	"log"
	"net/url"

	"github.com/gorilla/websocket"
)

func main() {
	// Define the WebSocket server URL
	serverURL := "ws://localhost:8080/ws"

	// Parse the URL
	u, err := url.Parse(serverURL)
	if err != nil {
		log.Fatal(err)
	}

	// Establish a WebSocket connection
	conn, _, err := websocket.DefaultDialer.Dial(u.String(), nil)
	if err != nil {
		log.Fatal(err)
	}
	defer conn.Close()

	fmt.Println("Connected to WebSocket server")

	// Start a goroutine to read messages from the server
	go readMessages(conn)

	// Send a message to the server
	message := "Hello, Server!"
	err = conn.WriteMessage(websocket.TextMessage, []byte(message))
	if err != nil {
		log.Println("Error sending message:", err)
		return
	}

	// Keep the program running
	select {}
}

func readMessages(conn *websocket.Conn) {
	for {
		messageType, msg, err := conn.ReadMessage()
		if err != nil {
			log.Println("Error reading message:", err)
			return
		}
		if messageType == websocket.TextMessage {
			handleMessage(msg)
		}
	}
}

func handleMessage(msg []byte) {
	fmt.Println("Received message:", string(msg))
}
```

### Output

1. Client side

```terminal
PS C:\Users\Dell\OneDrive\Documents\GitHub\golang-socketio-client> go run .\go-client.go
Connected to WebSocket server
Received message: Hello, Server!
```

2. Server side

```terminal
PS C:\Users\Dell\OneDrive\Documents\GitHub\golang-socketio-client> go run .\go-server.go
WebSocket server is running on ws://localhost:8080
Client connected
```

Of course this is just a basic example. I will need to perhaps create a public repo so that others can contribute to the same / develop it further.

**Comment by zishang520** [comment]
Welcome to use: https://github.com/zishang520/socket.io

The client is expected to be completed by the end of this year, I hope I have enough time (

**Comment by doquangtan** [comment]
https://github.com/doquangtan/gofiber-socket.io go lang socket.io server 4.x, supported client 3.x, 4x

**Comment by CorrectRoadH** [comment]
> Welcome to use: https://github.com/zishang520/socket.io
> 
> The client is expected to be completed by the end of this year, I hope I have enough time (

Hope there is a client version. 🥺

**Comment by CorrectRoadH** [comment]
> Do need Golang client, too. And I'd like to contribute to it, as soon as I can.

Did you find it? I need golang client too.

**Comment by zishang520** [comment]
The beta version of engine.io-client has been completed:  
https://github.com/zishang520/engine.io-client-go  

The socket.io-client is still in development—almost there, almost there! (Let me go back and hug my adorable little baby first 😆)  
https://github.com/zishang520/socket.io-client-go
